File: .claude/hooks/session-start-worktree-if-main.mjs
#!/usr/bin/env node
// SessionStart: crée un worktree isolé unique si la session démarre sur main/master.
// Chaque session obtient un worktree frais avec un suffixe aléatoire — jamais de
// réutilisation d'un worktree existant (un worktree désynchronisé provoquerait des conflits).
// Le nettoyage reste manuel (`git worktree prune`).
import { execSync } from 'child_process';
import { existsSync } from 'fs';
import { randomBytes } from 'crypto';
import { fileURLToPath } from 'url';

function defaultExec(cmd) {
  try { return execSync(cmd, { encoding: 'utf8', timeout: 10_000 }).trim(); } catch { return ''; }
}

function defaultAddWorktree(path, branchName) {
  execSync(`git worktree add "${path}" -b "${branchName}"`, {
    encoding: 'utf8',
    timeout: 15_000,
    stdio: ['ignore', 'ignore', 'ignore'],
  });
}

function defaultRandom(len = 6) {
  return randomBytes(Math.ceil(len / 2)).toString('hex').slice(0, len);
}

export function run({
  exec = defaultExec,
  addWorktree = defaultAddWorktree,
  exists = existsSync,
  now = () => new Date(),
  random = defaultRandom,
} = {}) {
  const branch = exec('git branch --show-current') || exec('git rev-parse --abbrev-ref HEAD');
  if (!branch || !/^(main|master)$/.test(branch)) return null;

  const currentRoot = exec('git rev-parse --show-toplevel');
  if (!currentRoot) return null;

  // Ne pas agir si on est déjà dans un worktree secondaire
  const worktreeList = exec('git worktree list');
  const mainRoot = worktreeList.split('\n')[0]?.split(/\s+/)[0] ?? '';
  if (mainRoot !== currentRoot) return null;

  // Synchroniser main avec le remote avant de créer le worktree
  exec('git fetch --quiet origin main');
  exec('git merge --ff-only origin/main');

  // Nom unique par session : date + suffixe hex aléatoire (même format que Claude Code App)
  const date = now().toISOString().slice(0, 10).replace(/-/g, '');
  const suffix = random(6);
  const branchName = `claude/session-${date}-${suffix}`;
  const worktreePath = `${currentRoot}/.claude/worktrees/session-${date}-${suffix}`;

  try {
    addWorktree(worktreePath, branchName);
  } catch {
    return [
      `## ⚠️  Session démarrée sur \`main\``,
      `- Impossible de créer un worktree automatiquement.`,
      `- Créez manuellement un worktree ou une branche avant de modifier des fichiers.`,
    ].join('\n') + '\n';
  }

  if (!exists(worktreePath)) return null;

  return [
    `## Worktree isolé créé automatiquement`,
    `- Session démarrée sur \`main\` : un worktree unique a été créé pour cette session.`,
    `- **Chemin** : \`${worktreePath}\``,
    `- **Branche** : \`${branchName}\``,
    `- Travaillez dans ce worktree — évitez de modifier des fichiers dans le dépôt principal.`,
  ].join('\n') + '\n';
}

/* v8 ignore next 4 */
if (process.argv[1] === fileURLToPath(import.meta.url)) {
  const result = run();
  if (result) process.stdout.write(result);
}

File: .claude/hooks/session-start-worktree-if-main.py
#!/usr/bin/env python3
# @hookstack session-start-worktree-if-main
"""Creates a fresh isolated worktree when the session starts on main/master (SessionStart).
Each session gets a fresh worktree with a random suffix — never reuses an existing
worktree (a desynchronized worktree would cause conflicts). Cleanup stays manual.
"""
import os
import secrets
import subprocess
import sys
from datetime import datetime, timezone


def _default_exec(cmd):
    try:
        return subprocess.run(
            cmd, shell=True, capture_output=True, text=True, timeout=10, check=True
        ).stdout.strip()
    except Exception:
        return ""


def _default_add_worktree(path, branch_name):
    subprocess.run(
        f'git worktree add "{path}" -b "{branch_name}"',
        shell=True,
        timeout=15,
        check=True,
        stdout=subprocess.DEVNULL,
        stderr=subprocess.DEVNULL,
    )


def _default_random(length=6):
    return secrets.token_hex(length // 2 + 1)[:length]


def run(
    *,
    exec_cmd=None,
    add_worktree=None,
    exists=None,
    now=None,
    random=None,
):
    if exec_cmd is None:
        exec_cmd = _default_exec
    if add_worktree is None:
        add_worktree = _default_add_worktree
    if exists is None:
        exists = os.path.exists
    if now is None:
        now = lambda: datetime.now(timezone.utc)
    if random is None:
        random = _default_random

    branch = exec_cmd("git branch --show-current") or exec_cmd(
        "git rev-parse --abbrev-ref HEAD"
    )
    if not branch or branch not in ("main", "master"):
        return None

    current_root = exec_cmd("git rev-parse --show-toplevel")
    if not current_root:
        return None

    # Ne pas agir si on est déjà dans un worktree secondaire
    worktree_list = exec_cmd("git worktree list")
    lines = worktree_list.split("\n") if worktree_list else []
    main_root = lines[0].split()[0] if lines and lines[0] else ""
    if main_root != current_root:
        return None

    # Synchroniser main avec le remote avant de créer le worktree
    exec_cmd("git fetch --quiet origin main")
    exec_cmd("git merge --ff-only origin/main")

    # Nom unique par session : date + suffixe hex aléatoire
    date = now().isoformat().split("T")[0].replace("-", "")
    suffix = random(6)
    branch_name = f"claude/session-{date}-{suffix}"
    worktree_path = f"{current_root}/.claude/worktrees/session-{date}-{suffix}"

    try:
        add_worktree(worktree_path, branch_name)
    except Exception:
        return "\n".join(
            [
                "## ⚠️  Session démarrée sur `main`",
                "- Impossible de créer un worktree automatiquement.",
                "- Créez manuellement un worktree ou une branche avant de modifier des fichiers.",
            ]
        ) + "\n"

    if not exists(worktree_path):
        return None

    return "\n".join(
        [
            "## Worktree isolé créé automatiquement",
            "- Session démarrée sur `main` : un worktree unique a été créé pour cette session.",
            f"- **Chemin** : `{worktree_path}`",
            f"- **Branche** : `{branch_name}`",
            "- Travaillez dans ce worktree — évitez de modifier des fichiers dans le dépôt principal.",
        ]
    ) + "\n"


if __name__ == "__main__":
    result = run()
    if result:
        sys.stdout.write(result)
